allium/lib/Parser/Parser.cpp

```diff
@@ -44,56 +45,56 @@ Optional<PredicateDecl> Parser::parsePredicateDecl() {
         Parameter param;
 
         do {
-            if(parseParameter().unwrapInto(param)) {
+            if(parseParameter().unwrapResultInto(param, errors)) {
                 parameters.push_back(param);
             } else {
                 if (parameters.size() == 0) {
-                    emitSyntaxError("Parentheses must not appear after predicate name for predicates with zero arguments.");
+                    errors.push_back(SyntaxError("Parentheses must not appear after predicate name for predicates with zero arguments.", lexer.peek_next().location));
                 } else {
-                    emitSyntaxError("Expected an additional argument after \",\" in parameter list.");
+                    errors.push_back(SyntaxError("Expected an additional parameter after \",\" in parameter list.", lexer.peek_next().location));
                 }
-
-                return rewindAndReturn();
             }
         } while(lexer.take(Token::Type::comma));
 
         if(!lexer.take(Token::Type::paren_r)) {
-            emitSyntaxError("Expected a \",\" or \")\" after parameter.");
-            return rewindAndReturn();
+            errors.push_back(SyntaxError("Expected a \",\" or \")\" after parameter.", lexer.peek_next().location));
         }
 
         std::vector<EffectRef> effects;
-        if(parseEffectList().unwrapGuard(effects)) {
+        if(parseEffectList().unwrapResultGuard(effects, errors)) {
             rewindAndReturn();
         }
 
-        return PredicateDecl(
-            Name<Predicate>(identifier.text),
-            parameters,
-            effects,
-            identifier.location);
+        return ParserResult<PredicateDecl>(
+            Optional(PredicateDecl(Name<Predicate>(identifier.text), parameters, effects, identifier.location)),
+            errors);
```

Comment:
I like that the parsers now only have a single return statement at the end! Accumulating errors by rewinding and falling through is elegant. Excellent work 😸